### ISSUE TITLE
macOS Catalina 10.15 search tag appear "E432: Tags file not sorted"

### ISSUE THREAD

**Issue by griffinqiu** [vim/vim#5039]
I have install vim by brew in macOS Catalina 10.15.

When I search or jump any tag. "E432: Tags file not sorted" appeared.

Even use vim help, the error appeared  such as:
:h \<c-w\>

VIM - Vi IMproved 8.1 (2018 May 18, compiled Sep  5 2019 21:03:34)

**Comment by hovsater** [comment]
I'm seeing the exact same problem with a setup that worked flawlessly for over ten years. I'm trying to lookup documentation using `:help <keyword>` and are instantly greeted with the error message.

**Command**

```
:help surround
```

**Output**

```
E432: Tags file not sorted: /Users/kevinsjoberg/.vim/plugged/lightline.vim/doc/tags
E432: Tags file not sorted: /Users/kevinsjoberg/.vim/plugged/vim-dirvish/doc/tags
E432: Tags file not sorted: /Users/kevinsjoberg/.vim/plugged/vim-abolish/doc/tags
E432: Tags file not sorted: /Users/kevinsjoberg/.vim/plugged/vim-commentary/doc/tags
E432: Tags file not sorted: /Users/kevinsjoberg/.vim/plugged/vim-dispatch/doc/tags
E432: Tags file not sorted: /Users/kevinsjoberg/.vim/plugged/vim-rails/doc/tags
E432: Tags file not sorted: /usr/local/share/vim/vim81/doc/tags
E426: tag not found: surround@en
```

**Vim**

```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Sep 30 2019 01:26:53)
macOS version
Included patches: 1-2100
Compiled by Homebrew
```

**Comment by eirnym** [comment]
For plugins it seems `tags` file should be sorted case insensitive, while sorting tags from Vim runtime files doesn't help at all

**Comment by andymass** [comment]
Did you try `set notagbsearch`.   If you want to work with unsorted tag files your file should start with

       !_TAG_FILE_SORTED	0	/comment/

**Comment by hovsater** [comment]
@andymass while I appreciate workarounds, I much rather fix the underlying issue.

**Comment by andymass** [comment]
@KevinSjoberg How did you generate these tag files?

**Comment by hovsater** [comment]
@andymass like I always do. Either using `:helptags ALL` or automatically using [vim-plug](https://github.com/junegunn/vim-plug). This has been working for close to ten years. The only difference is the recent upgrade to Mac OS Catalina.

**Comment by andymass** [comment]
Either vim is generating out of order tag files or its calculation of when a tag file is not sorted is wrong.  

https://github.com/neovim/neovim/issues/11196

**Comment by chrisbra** [comment]
could that be related to case insensitivity of the underlying filesystem?

**Comment by yegappan** [comment]
Back in 2003, I saw the following e-mail thread that discussed a
solution to this problem.

- Yegappan

==================================================
Here is the scoop.. Setting the env. var. LC_ALL to C should help
you build a sorted tag database and allow you tag lookups from the
vi/vim editors.

>
> What would cause a tags database to _not_ be sorted?
>

Given that your example shows a sort into lexical order instead of native
ascii order, I suspect the problem is recent addition of support for
locales.  This changes the semantics of sort(1)so that it sorts
lexically by default instead of ascii order.This was an incredibly
boneheaded decision (to change the default behavior of a tool), but it
seems to be fairly universal now.

Try setting unix environment LC_ALL to C, and see if generating tags
makes the problem go away.  BTW, so far I've only seen this class of
problems on linux; have they now shown up on solaris?

**Comment by eirnym** [comment]
> could that be related to case insensitivity of the underlying filesystem?

No, this wasn't a case before while filesystem is case insensitive for years.

**Comment by dm1try** [comment]
I posted the patch in https://github.com/neovim/neovim/issues/11196#issuecomment-541212889
feel free to apply

**Comment by romainl** [comment]
FWIW, under Mojave, doing `:%!sort -R` on `$VIMRUNTIME/doc/tags` or any plugin-local `tags` file doesn't break `:help foo` at all despite making the files pretty much unsorted.

This makes me think that this is not a matter of "sorted" versus "unsorted".

--- EDIT ---

Indeed, as evidenced by dm1try's patch, it looks like a filesystem issue.

**Comment by eirnym** [comment]
@dm1try does this patch works in others versions of macOS?

**Comment by dm1try** [comment]
@eirnym sure

**Comment by eirnym** [comment]
Then it looks fine for me.

**Comment by hovsater** [comment]
Addressed in Neovim (see https://github.com/neovim/neovim/pull/11222). Could we apply the same patch in Vim?

**Comment by nickshatilo** [comment]
I would appreciate it if anyone could look into this.

**Comment by chdiza** [comment]
Has anyone reported this as a bug to Apple?  If it can be reproduced with Catalina's system `vim` there's a better shot that they'll fix it.

**Comment by johnlamb** [comment]
The bug is reported to apple (https://github.com/neovim/neovim/issues/11196#issuecomment-541365466)
I have an open pull request with the same fix as Neovim.

**Comment by k-takata** [comment]
Fixed by #5061.
Closing.

**Comment by spencerdcarlson** [comment]
Fixed by [PR 955](https://github.com/StephenWall/vim/pull/955) in [v8.2.0314](https://github.com/vim/vim/releases/tag/v8.2.0314) original issue [5061](https://github.com/vim/vim/pull/5061)
